framework/di/Container.php

```diff
@@ -524,7 +525,7 @@ public function resolveCallableDependencies(callable $callback, $params = [])
                     unset($params[$name]);
                 } elseif (!$associative && isset($params[0]) && $params[0] instanceof $className) {
                     $args[] = array_shift($params);
-                } elseif (Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
+                } elseif (Yii::$app instanceof ServiceLocator && Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
```

Comment:
How many...
````
(unset)Isset x 1000000: 0.087s
(unset)Instanceof x 1000000: 0.089s
(set)Isset x 1000000: 0.144s
(set)Instanceof x 1000000: 0.069s
```
Basically the difference is irrelevant when `Yii::$app` is not set; however that's not the generic case. So when it's set, the code about twice as slow.
You mention it is called a lot; however from my testing it is not called much, since this is DI. If you are creating 100k+ objects using DI then generally you will have more to worry about than these 2ms...

Above test data was created using the following script:
````
<?php

include 'framework/Yii.php';
$start = microtime(true);
$repeat = 100000;
for ($i = 0; $i < $repeat; $i++) {
    isset(Yii::$app);
}
$end = microtime(true);
echo "(unset)Isset x $repeat: " . number_format($end - $start, 3) . "s\n";


$start = microtime(true);
for ($i = 0; $i < $repeat; $i++) {
    Yii::$app instanceof \yii\di\ServiceLocator;
}
$end = microtime(true);
echo "(unset)Instanceof x $repeat: " . number_format($end - $start, 3) . "s\n";

Yii::$app = new \yii\di\ServiceLocator();

for ($i = 0; $i < $repeat; $i++) {
    isset(Yii::$app);
}
$end = microtime(true);
echo "(set)Isset x $repeat: " . number_format($end - $start, 3) . "s\n";


$start = microtime(true);
for ($i = 0; $i < $repeat; $i++) {
    Yii::$app instanceof \yii\di\ServiceLocator;
}
$end = microtime(true);
echo "(set)Instanceof x $repeat: " . number_format($end - $start, 3) . "s\n";
````

I strongly feel that in this case the argument for bad code in favor of perceived performance is not valid.